In [ ]:
# Simplified code demo of LSA

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

# Sample common text corpus
documents = [
    "The cat sat on the mat.",
    "Dogs are great companions.",
    "The sun is bright today.",
    "Cats and dogs are popular pets.",
    "The weather is nice for a walk."
]

# Step 1: Data Cleaning and Formatting for Analysis,
# Skipping this step as the data has already been cleaned.

# Step 2: Construct the TF-IDF matrix
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documents)

# Step 3: Apply SVD to reduce dimensions
svd = TruncatedSVD(n_components=2)
lsa_matrix = svd.fit_transform(tfidf_matrix)

# Step 4: Analyze the results
# Create a DataFrame for viewing
lsa_df = pd.DataFrame(lsa_matrix, columns=['Concept 1', 'Concept 2'])
print("Reduced Dimensions (Latent Concepts):")
print(lsa_df)

Reduced Dimensions (Latent Concepts):
      Concept 1     Concept 2
0  6.263182e-01  1.932910e-16
1 -4.140572e-16  8.096747e-01
2  7.173435e-01 -7.926662e-15
3 -1.891549e-16  8.096747e-01
4  6.985829e-01  8.989044e-15


- Below is the interpretation of the above output matris.


- In the output matrix, rows represent individual documents.
- Columns represent latent concepts or themes. 
- Values indicate the strength of each document's relationship to those concepts.
- The first and third documents (column 1) show high values. It shows strong relevance to Concept 1.
- The second document has a significant value (column 2). It shows strong relevance to Concept 1.
- Other documents in both columns are less relevant.
- The first and third documents may share a theme.
- The second document relates to a different topic.
- Higher scores indicate thematic similarities among documents. It helps in clustering and topic identification.

Code Snippet 7.2